File: pintia/gplt_practice/l2/_020.java
import java.io.*;
import java.util.*;

public class _020 {
    static final int N = (int) 1e5 + 10;
    static int n;
    static double z, r;
    static int[] scale = new int[N];
    static int[] h = new int[N], e = new int[N], ne = new int[N];
    static int idx = 0;

    static void add(int a, int b) {
        e[idx] = b;
        ne[idx] = h[a];
        h[a] = idx++;
    }

    static double dfs(int u, double x) {
        if (scale[u] != 0) return x * r * scale[u];
        double w = x * r;
        double sum = 0.0;
        for (int i = h[u]; i != -1; i = ne[i]) {
            int v = e[i];
            sum += dfs(v, w);
        }
        return sum;
    }

    public static void main(String[] args) throws IOException {
        StreamTokenizer in = new StreamTokenizer(new BufferedReader(new InputStreamReader(System.in)));
        in.nextToken(); n = (int) in.nval;
        in.nextToken(); z = in.nval;
        in.nextToken(); r = in.nval;
        r = (100.0 - r) / 100.0;
        Arrays.fill(h, -1);
        int k, v;
        for (int u = 0; u < n; ++u) {
            in.nextToken(); k = (int) in.nval;
            if (k == 0) {
                in.nextToken(); scale[u] = (int) in.nval;
                continue;
            }
            while (k-- > 0) {
                in.nextToken(); v = (int) in.nval;
                add(u, v);
            }
        }
        if (n == 1) {
            System.out.printf("%.0f\n", z * scale[0] - 0.5);
            return;
        }
        double sum = 0.0;
        for (int i = h[0]; i != -1; i = ne[i]) sum += dfs(e[i], z);
        System.out.printf("%.0f\n", sum - 0.5);
    }
}

File: pintia/gplt_practice/l2/_020.cpp
#include <cstdio>
#include <cstring>

using namespace std;

typedef long long LL;

const int N = 1e5 + 10;
int n;
double z, r;
int scale[N];
int h[N], e[N], ne[N], idx;

void add(int a, int b) {
    e[idx] = b, ne[idx] = h[a], h[a] = idx++;
}

double dfs(int u, double x) {
    if (scale[u]) return x * r * scale[u];
    double w = x * r;
    double sum = 0.0;
    for (int i = h[u]; ~i; i = ne[i]) {
        int v = e[i];
        sum += dfs(v, w);
    }
    return sum;
}

int main() {
    scanf("%d%lf%lf", &n, &z, &r);
    r = (100 - r) / 100;
    memset(h, -1, sizeof(h));
    int k, v;
    for (int u = 0; u < n; ++u) {
        scanf("%d", &k);
        if (!k) {
            scanf("%d", &scale[u]);
            continue;
        }
        while (k--) {
            scanf("%d", &v);
            add(u, v);
        }
    }
    if (r == 0.0) {
        printf("%lld\n", (LL) (z * scale[0]));
        return 0;
    }
    printf("%lld\n", (LL) dfs(0, z / r));

    return 0;
}